### PR TITLE
Polars: add cut transformation and categorical dtype

### LINKED ISSUE #894: Polars: add `make_expr_cut`
**Description**
- **Constructor**: Transformation
- **Class**: Expr
- **DSL Variant:** Function: Cut 
   - **Purpose**: for multi-quantiles

--- 
reference: [Polars APIs sheet](https://docs.google.com/spreadsheets/d/1w4H-qrSe9pkvNT2NhowfiIHNWXSseOyMX2nu5H6-RkQ/edit?gid=0#gid=0)

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #894

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#2004** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2004?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2003** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2003?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/domains/polars/series/mod.rs [hunk 1]
```diff
@@ -215,12 +215,64 @@ impl<T: CheckAtom + PrimitiveDataType> SeriesAtomDomain for OptionDomain<AtomDom
...
+}
+
+impl CategoricalDomain {
+    /// Only use this constructor if you know both the category set,
+    /// as well as how the categories are encoded.
+    ///
+    /// Typically when Polars constructs categorical data,
+    /// it assigns indices by the order encountered in the data, making the encoding data-dependent.
+    /// An example where this can be called is for categorical data emitted by the Polars cut expression,
+    /// where the categories and encoding are pre-determined by the expression (the bin edges).
```
Comment (mccalluc) [inline rust/src/domains/polars/series/mod.rs:231]: This is interesting. I had tried to use category at some point in a docs example, for something like state abbreviations, but it sounds like we would never recommend that. (But this isn't bootstrapped, so it's not showing up in the generated code in any case.)

Maybe the actionable thing here is to note in our Polars table that categories aren't available, and explain why not? File an issue from this comment if that seems like a good route.

**Review by mccalluc — DISMISSED** [review]
Looks good. Think I've reviewed bits of this before, though not in this PR? Let me know if anything deserves a closer look.

File: rust/src/domains/polars/series/mod.rs (hunk 1 above)
Comment (Shoeboxam) [reply rust/src/domains/polars/series/mod.rs]: I've gone ahead and fleshed this PR out to actually add categoricals. You can use categorical data with unknown encodings, so long as you don't read or change the encoding.

File: python/test/test_polars.py [hunk 2]
```diff
@@ -692,4 +692,52 @@ def test_csv_bad_encoding_loading():
...
+
+    lf_domain = dp.lazyframe_domain([dp.series_domain("A", dp.categorical_domain())])
+    assert str(lf_domain) == "FrameDomain(A: cat; margins=[])"
+
+    lf_domain = dp.lazyframe_domain([dp.series_domain("A", dp.option_domain(dp.categorical_domain()))])
+    assert str(lf_domain) == "FrameDomain(A: cat; margins=[])"
+
+    lf = pl.LazyFrame([pl.Series("A", ["Texas", "New York", None], dtype=pl.Categorical)])
+
+    with pytest.raises(dp.OpenDPException):
```
Comment (mccalluc) [inline python/test/test_polars.py:709]: Also check the error message? Want to be sure that it makes sense to the user.

(I also see elsewhere in this PR that the categorical is wrapped by optional... but I guess that doesn't mean we can fill it with nulls? -- If that does enable anything, special that too.)

**Review by mccalluc — COMMENTED** [review]
Thanks for adding Categoricals. I think #2025 can cover any docs.

Where we have a `pytest.raises`, I'd have a preference to also check the text of the error message, to make sure we're hitting the error we think we should hit, and to encourage us to think about the usefulness of the error to end users.

(I wouldn't make this a rule on every project: if the error classes are fine grained, that is enough, but we only have a few, and we know that the embedded Rust errors make it harder to understand, so I think we need an extra checks to make sure it's informative.)

(But if that bar is too high, and it would discourage tests at least this good, please ignore.)

File: python/test/test_polars.py (hunk 2 above)
Comment (Shoeboxam) [reply python/test/test_polars.py:709]: I explicitly check both now.

**Review by mccalluc — APPROVED** [review]
Thanks!

**Pull request merged**